src/store/EncryptedStore.ts

```diff
@@ -30,39 +34,119 @@ export default class EncryptedStore implements KeyStore {
     const keyAddress = await this.getStorageAddress(KEY_BUNDLE_NAME)
-    const encodedBundle = await bundle.toEncryptedBytes(this.signer)
+    const encodedBundle = await this.toEncryptedBytes(bundle, this.signer)
     // We need to setup the Authenticator so that the underlying store can publish messages without error
     if (typeof this.store.setAuthenticator === 'function') {
       this.store.setAuthenticator(new Authenticator(bundle.identityKey))
     }
     await this.store.set(keyAddress, Buffer.from(encodedBundle))
   }
+
+  // encrypts/serializes the bundle for storage
+  async toEncryptedBytes(
+    bundle: PrivateKeyBundleV1,
+    wallet: Signer
+  ): Promise<Uint8Array> {
+    // serialize the contents
+    const bytes = bundle.encode()
+    const wPreKey = getRandomValues(new Uint8Array(32))
+    const secret = hexToBytes(
+      await wallet.signMessage(storageSigRequestText(wPreKey))
+    )
+    const ciphertext = await encrypt(bytes, secret)
+    return proto.EncryptedPrivateKeyBundle.encode({
+      v1: {
+        walletPreKey: wPreKey,
+        ciphertext,
+      },
+    }).finish()
+  }
+
+  // decrypts/deserializes the bundle from storage bytes
+  async fromEncryptedBytes(
+    wallet: Signer,
+    bytes: Uint8Array
+  ): Promise<[PrivateKeyBundleV1, boolean]> {
+    const [eBundle, needsUpdate] = getEncryptedBundle(bytes)
+
+    if (!eBundle.walletPreKey) {
+      throw new Error('missing wallet pre-key')
+    }
+    if (!eBundle.ciphertext?.aes256GcmHkdfSha256) {
+      throw new Error('missing bundle ciphertext')
+    }
+
+    const secret = hexToBytes(
+      await wallet.signMessage(storageSigRequestText(eBundle.walletPreKey))
+    )
+    const ciphertext = new Ciphertext(eBundle.ciphertext)
+    const decrypted = await decrypt(ciphertext, secret)
+    const [bundle, needsUpdate2] = getPrivateBundle(decrypted)
+    return [bundle, needsUpdate || needsUpdate2]
+  }
+}
+
+// getEncryptedV1Bundle returns the decoded bundle from the provided bytes. If there is an error decoding the bundle it attempts
+// to decode the bundle as a legacy bundle. Additionally return whether the bundle is in the expected format.
+function getEncryptedBundle(
+  bytes: Uint8Array
+): [proto.EncryptedPrivateKeyBundleV1, boolean] {
+  try {
+    const b = proto.EncryptedPrivateKeyBundle.decode(bytes)
+    if (b.v1) {
+      return [b.v1, false]
+    }
+  } catch (e) {
+    return [proto.EncryptedPrivateKeyBundleV1.decode(bytes), true]
+  }
+  throw new Error('unrecognized encrypted private key bundle version')
+}
+
+// getPrivateV1Bundle returns the decoded bundle from the provided bytes. If there is an error decoding the bundle it attempts
+// to decode the bundle as a legacy bundle. Additionally return whether the bundle is in the expected format.
+function getPrivateBundle(bytes: Uint8Array): [PrivateKeyBundleV1, boolean] {
+  try {
+    const b = DecodePrivateKeyBundle(bytes) as PrivateKeyBundleV1
```

Comment:
This type assertion seems dangerous, given the function can return either V1 or V2 bundles. Today, we only write V1 so it's safe in practice. But we'll need to be careful that we update this logic to actually check the return value once we have the possibility of V1 or V2 bundles being stored on the network.

Reply:
All these assertions are temporary and should be removed as we roll out V2. I can mark them with TODO comments if that helps.